File: apps/orchestrator-go/internal/template/template.go
// Package template defines the Go representation of a rule template.
// The single source of truth is packages/rule-schema/schema/rule-template.schema.json;
// this file is kept in lockstep via the golden fixture round-trip test.
package template

import "encoding/json"

type SourceField struct {
	Name string `json:"name"`
	Type string `json:"type"`
}

type Source struct {
	Type        string                 `json:"type"`
	Schema      []SourceField          `json:"schema,omitempty"`
	Options     map[string]any         `json:"options,omitempty"`
	ConnectorID *int64                 `json:"connectorId,omitempty"`
}

type PreprocessStep struct {
	ID    string         `json:"id"`
	Field string         `json:"field"`
	Fn    string         `json:"fn"`
	Args  map[string]any `json:"args,omitempty"`
	Code  string         `json:"code,omitempty"`
}

type Auth struct {
	Type       string `json:"type"`
	SecretRef  string `json:"secretRef,omitempty"`
	HeaderName string `json:"headerName,omitempty"`
}

// Destination is the HTTP target for each row. URL/path/query/headers may be
// templated per row using the same `{"$from":"field"}` / `{"$literal":"..."}`
// sigils as the JSON body mapping.
type Destination struct {
	Type              string             `json:"type"`
	Method            string             `json:"method"`
	URL               string             `json:"url"`
	PathParams        map[string]any     `json:"pathParams,omitempty"`
	QueryParams       map[string]any     `json:"queryParams,omitempty"`
	Headers           map[string]any     `json:"headers,omitempty"`
	Auth              *Auth              `json:"auth,omitempty"`
	IdempotencyKey    json.RawMessage    `json:"idempotencyKey,omitempty"`
	IdempotencyHeader string             `json:"idempotencyHeader,omitempty"`
}

type Retry struct {
	MaxAttempts       int    `json:"maxAttempts,omitempty"`
	Backoff           string `json:"backoff,omitempty"`
	InitialIntervalMs int    `json:"initialIntervalMs,omitempty"`
}

// Step is one HTTP call in a multi-step chain. Later steps may reference
// earlier responses via {"$fromResponse":"stepName","path":"$.jsonpath"}.
type Step struct {
	Name        string         `json:"name"`
	Description string         `json:"description,omitempty"`
	Mapping     map[string]any `json:"mapping,omitempty"`
	Destination Destination    `json:"destination"`
	// OnFailure controls chain behavior after this step definitively fails
	// (4xx or retries exhausted). "" or "stop" (default) ends the chain;
	// "continue" runs later steps. The row is still marked failed if any
	// step failed.
	OnFailure string `json:"onFailure,omitempty"`
}

// ContinuesOnFailure reports whether later steps should run after this step fails.
func (s *Step) ContinuesOnFailure() bool {
	return s.OnFailure == "continue"
}

// PayloadTemplate returns the step's payload mapping (may be nil for
// payload-less calls such as GET).
func (s *Step) PayloadTemplate() map[string]any {
	if s.Mapping == nil {
		return nil
	}
	if m, ok := s.Mapping["payload"].(map[string]any); ok {
		return m
	}
	return nil
}

type RuleTemplate struct {
	ID          string                 `json:"id"`
	Version     int                    `json:"version"`
	Name        string                 `json:"name"`
	Source      Source                 `json:"source"`
	Preprocess  []PreprocessStep       `json:"preprocess"`
	Mapping     map[string]any         `json:"mapping,omitempty"`
	Destination Destination            `json:"destination,omitempty"`
	Steps       []Step                 `json:"steps,omitempty"`
	Retry       *Retry                 `json:"retry,omitempty"`
	Concurrency int                    `json:"concurrency,omitempty"`
}

func (rt *RuleTemplate) PayloadTemplate() map[string]any {
	if m, ok := rt.Mapping["payload"].(map[string]any); ok {
		return m
	}
	return nil
}

// ExecutionSteps normalizes either template shape into an ordered step list.
// Single-destination templates become a one-step chain named "main".
func (rt *RuleTemplate) ExecutionSteps() []Step {
	if len(rt.Steps) > 0 {
		return rt.Steps
	}
	return []Step{{Name: "main", Mapping: rt.Mapping, Destination: rt.Destination}}
}

File: apps/transform-worker/migration_worker/template.py
"""Pydantic models matching the shared rule-template JSON Schema."""
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class SourceField(BaseModel):
    model_config = ConfigDict(extra="forbid")
    name: str
    type: Literal["string", "integer", "number", "boolean", "datetime", "object"]


class Source(BaseModel):
    model_config = ConfigDict(extra="forbid")
    type: Literal["csv", "json", "xml", "salesforce", "watched_prefix"]
    schema_: list[SourceField] | None = Field(default=None, alias="schema")
    options: dict[str, Any] | None = None
    connectorId: int | None = None


class PreprocessStep(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: str
    field: str
    fn: str
    args: dict[str, Any] | None = None
    code: str | None = None


class Auth(BaseModel):
    type: Literal["bearer", "basic", "api_key", "oauth2_client"]
    secretRef: str | None = None
    headerName: str | None = None


class Destination(BaseModel):
    model_config = ConfigDict(extra="forbid")
    type: Literal["http"]
    method: Literal["GET", "POST", "PUT", "PATCH", "DELETE"]
    # Plain str (not HttpUrl): URLs may contain {placeholder} path templating.
    url: str
    pathParams: dict[str, Any] | None = None
    queryParams: dict[str, Any] | None = None
    headers: dict[str, Any] | None = None
    auth: Auth | None = None
    idempotencyKey: Any | None = None
    idempotencyHeader: str | None = None


class Retry(BaseModel):
    model_config = ConfigDict(extra="forbid")
    maxAttempts: int | None = Field(default=None, ge=1, le=20)
    backoff: Literal["exponential", "fixed"] | None = None
    initialIntervalMs: int | None = Field(default=None, ge=10)


class Mapping(BaseModel):
    model_config = ConfigDict(extra="forbid")
    payload: dict[str, Any]


class Step(BaseModel):
    """One HTTP call in a multi-step chain."""

    model_config = ConfigDict(extra="forbid")
    name: str = Field(pattern=r"^[A-Za-z_][A-Za-z0-9_]*$")
    description: str | None = None
    mapping: Mapping | None = None
    destination: Destination
    onFailure: Literal["stop", "continue"] | None = None


class ExportColumn(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)
    name: str
    from_field: str | None = Field(default=None, alias="$from")
    from_response: str | None = Field(default=None, alias="$fromResponse")
    path: str | None = None


class ExportSpec(BaseModel):
    model_config = ConfigDict(extra="forbid")
    columns: list[ExportColumn] | None = None


class RuleTemplate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    id: str
    version: int = Field(ge=1)
    name: str
    source: Source
    preprocess: list[PreprocessStep]
    # Exactly one of (mapping + destination) or steps -- enforced below.
    mapping: Mapping | None = None
    destination: Destination | None = None
    steps: list[Step] | None = Field(default=None, min_length=1, max_length=10)
    retry: Retry | None = None
    concurrency: int | None = Field(default=None, ge=1, le=512)
    export: ExportSpec | None = None

    @model_validator(mode="after")
    def _one_shape(self) -> "RuleTemplate":
        has_single = self.mapping is not None and self.destination is not None
        has_steps = self.steps is not None
        if has_single == has_steps:
            raise ValueError(
                "template must have either mapping+destination or steps, not both/neither"
            )
        return self

    def execution_steps(self) -> list[Step]:
        """Normalize either template shape into an ordered step list."""
        if self.steps is not None:
            return self.steps
        assert self.mapping is not None and self.destination is not None
        return [Step(name="main", mapping=self.mapping, destination=self.destination)]
